pkg/typesystem/typesystem.go

```diff
@@ -379,6 +387,59 @@ func (t *TypeSystem) UsersetCanFastPath(relationReferences []*openfgav1.Relation
 	return true
 }
 
+// recursiveUsersetNodeCanFastpath is a helper function to determine whether the node (object#relation) has
+// - one edge back to itself
+// - other edges lead directly to terminal types (not union/intersection/exclusion).
+func (t *TypeSystem) recursiveUsersetNodeCanFastpath(curAuthorizationModelNode *graph.AuthorizationModelNode, userType string) bool {
+	hasNodeBackToItself := false
+	hasCorrectTerminalType := false
+	childrenNodes := t.authorizationModelGraph.To(curAuthorizationModelNode.ID())
+	for childrenNodes.Next() {
+		curNode := childrenNodes.Node()
+		curChildAuthorizationModelNode, ok := curNode.(*graph.AuthorizationModelNode)
+		if !ok {
+			// Strange, I would have expected to be able to cast to AuthorizationModelNode
+			return false
+		}
+		childNodeObjectType, childNodeRelation := tuple.SplitObjectRelation(curChildAuthorizationModelNode.Label())
+		if childNodeRelation == "" {
+			// this means the childNode is either a terminal type or one of relation keyword
+			if childNodeObjectType == userType {
+				hasCorrectTerminalType = true
+			} else if childNodeObjectType == "union" || childNodeObjectType == "intersection" || childNodeObjectType == "exclusion" {
```

Comment:
This could perhaps be expressed as a set (`map[string]struct`) named setOperators and checking if its in the map.

Reply:
change to switch/case